source/extensions/filters/network/mongo_proxy/proxy.cc

```diff
@@ -58,9 +80,39 @@ ProxyFilter::ProxyFilter(const std::string& stat_prefix, Stats::Scope& scope,
 
 ProxyFilter::~ProxyFilter() { ASSERT(!delay_timer_); }
 
+void ProxyFilter::setDynamicMetadata(std::string op, absl::optional<std::string> db,
+                                     absl::optional<std::string> collection) {
+  ProtobufWkt::Struct metadata(
+      (*read_callbacks_->connection()
+            .streamInfo()
+            .dynamicMetadata()
+            .mutable_filter_metadata())[NetworkFilterNames::get().MongoProxy]);
+  auto& fields = *metadata.mutable_fields();
+  auto& list = *fields[DynamicMetadataKeysSingleton::get().MessagesField].mutable_list_value();
+  auto& message = *list.add_values()->mutable_struct_value()->mutable_fields();
+
+  message[DynamicMetadataKeysSingleton::get().OperationField].set_string_value(op);
+  if (db) {
+    message[DynamicMetadataKeysSingleton::get().DatabaseField].set_string_value(db.value());
+  }
+  if (collection) {
+    message[DynamicMetadataKeysSingleton::get().CollectionField].set_string_value(
+        collection.value());
+  }
+
+  read_callbacks_->connection().streamInfo().setDynamicMetadata(
+      NetworkFilterNames::get().MongoProxy, metadata);
+}
+
 void ProxyFilter::decodeGetMore(GetMoreMessagePtr&& message) {
   tryInjectDelay();
 
+  if (emit_dynamic_metadata_) {
+    std::vector<std::string> names = absl::StrSplit(message->fullCollectionName(), '.');
+    setDynamicMetadata(DynamicMetadataKeysSingleton::get().OperationGetMore,
+                       absl::make_optional(names[0]), absl::make_optional(names[1]));
+  }
```

Comment:
If GetMore cannot happen on its own (i.e. there has to be a Get beofre GetMore), then we can drop this as well.

Reply:
Addressed all comments in 31c55e7.